File: CSharpPools/PoolsLibrary/Pool/BasicFunctionality/PoolItemsStorage.cs
using System;
using System.Collections.Concurrent;
using System.Threading.Tasks;
using PoolsLibrary.ObjectActions;
using PoolsLibrary.ObjectUtilization;
using PoolsLibrary.Pool.BasicFunctionality.Item;
using PoolsLibrary.Pool.BasicFunctionality.Storages;

namespace PoolsLibrary.Pool.BasicFunctionality
{
    public class PoolItemsStorage<TK, TV> : IInternalPool<TK, TV>
    {
        ConcurrentDictionary<TK, PoolItem<TK, TV>> IInternalPool<TK, TV>.KeyToPoolItem
        {
            get { return _keyToPoolItem; }
        }

        private readonly PoolItemsStorageSettings _settings;
        private readonly ConcurrentDictionary<TK, PoolItem<TK, TV>> _keyToPoolItem;

        private readonly IPoolObjectActions<TV> _objectActions;
        private readonly IObjectUtilizer<TK, TV> _objectUtilizer;

        public PoolItemsStorage(PoolItemsStorageSettings settings,
                        IPoolObjectActions<TV> objectActions,
                        IObjectUtilizer<TK, TV> objectUtilizer)
        {
            _settings = settings;
            _objectActions = objectActions;
            _objectUtilizer = objectUtilizer;

            _keyToPoolItem = new ConcurrentDictionary<TK, PoolItem<TK, TV>>();
            _objectUtilizer.ObjectIsGone += OnObjectIsGoneEvent;
        }

        public bool TryObtain(TK key, out TV outPoolObject, Func<TK, TV> createDelegateIfNoObjectIsAvailable)
        {
            var poolItem = _keyToPoolItem.GetOrAdd(key, CreatePoolItem);
            return poolItem.TryObtain(out outPoolObject, createDelegateIfNoObjectIsAvailable);
        }

        public void Release(TK key, TV objectToRelease)
        {
        }

        public void Dispose()
        {
            _objectUtilizer.ObjectIsGone -= OnObjectIsGoneEvent;
            Parallel.ForEach(_keyToPoolItem.Values, x => x.Dispose());
        }

        private void OnObjectIsGoneEvent(object sender, GoneObjectEventArgs<TK, TV> goneObjectArgs)
        {
            if (goneObjectArgs.Reporter.GetType() == typeof (PoolItem<TK, TV>))
                return;

            PoolItem<TK, TV> outPoolItem;
            if (_keyToPoolItem.TryGetValue(goneObjectArgs.Key, out outPoolItem))
                outPoolItem.MarkObjectForKilling(goneObjectArgs.PoolObject);
        }

        private PoolItem<TK, TV> CreatePoolItem(TK key)
        {
            var poolItemSettings = new PoolItemSettings<TK>
            {
                Key = key,
                MarkObtainedObjectAsNotAvailable = _settings.AllowOnlyOneUserPerObject,
                MaxObjectsCount = _settings.MaxObjectsCountPerKey,
                ThrowIfCantCreateNewBecauseOfReachedLimit = _settings.ThrowIfCantCreateObjectBecauseOfReachedLimit,
            };
            var availableObjectsStorage = StorageBase<TV>.Create(_settings.BalancingStrategy);

            return new PoolItem<TK, TV>(poolItemSettings, availableObjectsStorage, _objectActions, _objectUtilizer);
        }
    }
}

File: JavaPools/PoolsLibrary/src/main/java/ru/urfu/javapools/poolslibrary/pool/basicfunctionality/PoolItemsStorage.java
package ru.urfu.javapools.poolslibrary.pool.basicfunctionality;

import java.util.Collection;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;

import ru.urfu.javapools.poolslibrary.controller.InvalidPoolOperationException;
import ru.urfu.javapools.poolslibrary.controller.PoolObjectObtainingInterruptedException;
import ru.urfu.javapools.poolslibrary.function.FunctionThatMayThrow;
import ru.urfu.javapools.poolslibrary.objectactions.IPoolObjectActions;
import ru.urfu.javapools.poolslibrary.objectutilization.GoneObjectEvent;
import ru.urfu.javapools.poolslibrary.objectutilization.IObjectUtilizationListener;
import ru.urfu.javapools.poolslibrary.objectutilization.IObjectUtilizer;
import ru.urfu.javapools.poolslibrary.pool.IPool;
import ru.urfu.javapools.poolslibrary.pool.basicfunctionality.item.NoAvailableObjectException;
import ru.urfu.javapools.poolslibrary.pool.basicfunctionality.item.ObjectCreationFailedException;
import ru.urfu.javapools.poolslibrary.pool.basicfunctionality.item.ObjectsMaxCountReachedException;
import ru.urfu.javapools.poolslibrary.pool.basicfunctionality.item.PoolItem;
import ru.urfu.javapools.poolslibrary.pool.basicfunctionality.item.PoolItemSettings;

public class PoolItemsStorage<TK,TV> implements IPool<TK,TV>,
												IObjectUtilizationListener<TK,TV> {
	
	private final PoolItemsStorageSettings _settings;
	private final ConcurrentHashMap<TK,PoolItem<TK,TV>> _keyToPoolItem;
	
	private final IPoolObjectActions<TV> _objectActions;
    private final IObjectUtilizer<TK, TV> _objectUtilizer;
	
	public PoolItemsStorage (PoolItemsStorageSettings settings,
            				 IPoolObjectActions<TV> objectActions,
            				 IObjectUtilizer<TK, TV> objectUtilizer) {
		_settings = settings;
		_objectActions = objectActions;
		_objectUtilizer = objectUtilizer;
		
		_keyToPoolItem = new ConcurrentHashMap<TK, PoolItem<TK,TV>>();
		_objectUtilizer.addListener(this);
	}
	
	@Override
	public TV obtain(TK key, FunctionThatMayThrow<TK, TV> createDelegateIfNoObjectIsAvailable)
		throws NoAvailableObjectException, ObjectsMaxCountReachedException,
			   ObjectCreationFailedException, PoolObjectObtainingInterruptedException {
			
		PoolItem<TK,TV> poolItem = _keyToPoolItem.computeIfAbsent(key, this::createPoolItem);
		return poolItem.obtain(createDelegateIfNoObjectIsAvailable);
	}

	@Override
	public void release(TK key, TV objectToRelease)	throws InvalidPoolOperationException {
	}
	
	@Override
	public void onObjectUtilization(GoneObjectEvent<TK,TV> goneObjectEvent) {
		
		PoolItem<TK,TV> poolItem = _keyToPoolItem.get(goneObjectEvent.getKey());
		if (poolItem != null)
			poolItem.markObjectForKilling(goneObjectEvent.getPoolObject());
	}	
	
	@Override
	public void close()	throws Exception {
		
		_objectUtilizer.removeListener(this);
		
		for (PoolItem<TK,TV> poolItem : _keyToPoolItem.values())
			poolItem.close();
	
		//_keyToPoolItem.values().parallelStream().forEach(poolItem -> poolItem.close());
	}	

	@Override
	public ConcurrentMap<TK, PoolItem<TK, TV>> getKeyToPoolItem() {
		return _keyToPoolItem;
	}
	
    private PoolItem<TK, TV> createPoolItem(TK key) {
    	
    	PoolItemSettings<TK> poolItemSettings = new PoolItemSettings<TK>() {{
    		setKey(key);
    		setMarkObtainedObjectAsNotAvailable(_settings.getAllowOnlyOneUserPerObject());
    		setMaxObjectsCount(_settings.getMaxObjectsCountPerKey());
        }};
        
        Collection<TV> availableObjectsStorage = _settings.getBalancingStrategy().createStorage();

        return new PoolItem<TK, TV>(poolItemSettings, availableObjectsStorage, _objectActions);
    }
}
